pkg/operator/controllers/checkers/serviceprincipalchecker/controller_test.go

```diff
@@ -0,0 +1,132 @@
+package serviceprincipalchecker
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"testing"
+	"time"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	arofake "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned/fake"
+	checkercommon "github.com/Azure/ARO-RP/pkg/operator/controllers/checkers/common"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
+	"github.com/Azure/ARO-RP/pkg/util/cmp"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+)
+
+type fakeChecker func(ctx context.Context, AZEnvironment string) error
+
+func (fc fakeChecker) Check(ctx context.Context, AZEnvironment string) error {
+	return fc(ctx, AZEnvironment)
+}
```

Comment:
this pattern seems a little weird tbh. what does it achieve, is it a function on a function? not 100% sure I understand it (maybe just missing something)
shouldn't the checker be an interface if we need to mock it to test the reconciler?

Reply:
Answered in [the thread below](https://github.com/Azure/ARO-RP/pull/2588/files#r1048214246).